File: Python/openstk/gfx/egin/egin_animate.py
from __future__ import annotations
import quaternion as quat, numpy as np
from typing import NamedTuple
from enum import Enum
from openstk.gfx.util import _np_createFromQuaternion4x4, _np_createTranslation4x4

#region Animate

# bone
class Bone:
    index: int
    parent: Bone
    children: list[Bone] = []
    name: str
    position: np.ndarray #Vector3
    angle: quat.quaternion
    bindPose: np.ndarray #Matrix4x4
    inverseBindPose: np.ndarray #Matrix4x4

    def __init__(self, index: int, name: str, position: np.ndarray, rotation: quat.quaternion):
        self.index = index
        self.name = name
        self.position = position
        self.angle = rotation
        # calculate matrices
        self.bindPose = np.matmul(_np_createFromQuaternion4x4(rotation), _np_createTranslation4x4(position))
        self.inverseBindPose = np.linalg.inv(self.bindPose)

    def setParent(self, parent: Bone) -> None:
        if parent not in self.children:
            self.parent = parent
            parent.children.append(self)

# ISkeleton
class ISkeleton:
    roots: list[Bone]
    bones: list[Bone]

# ChannelAttribute
class ChannelAttribute(Enum):
    Position = 0
    Angle = 1
    Scale = 2
    Unknown = 3

# FrameBone
class FrameBone:
    position: np.ndarray #Vector3
    angle: quat.quaternion
    scale: float

# Frame
class Frame:
    bones: list[FrameBone]

    def __init__(self, skeleton: ISkeleton):
        self.bones = [FrameBone]*len(skeleton.bones)
        self.clear(skeleton)

    def setAttribute(self, bone: int, attribute: ChannelAttribute, data: np.ndarray | quat.quaternion | float) -> None:
        match data:
            case p if isinstance(data, np.ndarray):
                match attribute:
                    case ChannelAttribute.Position: self.bones[bone].position = p
#if DEBUG
                    case _: print(f"Unknown frame attribute '{attribute}' encountered with Vector3 data")
#endif
            case q if isinstance(data, quat.quaternion):
                match attribute:
                    case ChannelAttribute.Angle: self.bones[bone].angle = q
#if DEBUG
                    case _: print(f"Unknown frame attribute '{attribute}' encountered with Quaternion data")
#endif
            case f if isinstance(data, float):
                match attribute:
                    case ChannelAttribute.Scale: self.bones[bone].scale = f
#if DEBUG
                    case _: print(f"Unknown frame attribute '{attribute}' encountered with float data")
#endif
            case _: raise Exception(f'Unknown {data}')

    def clear(self, skeleton: ISkeleton) -> None:
        for i in range(len(self.bones)):
            self.bones[i].position = skeleton.bones[i].position
            self.bones[i].angle = skeleton.bones[i].angle
            self.bones[i].scale = 1.

# IAnimation
class IAnimation:
    name: str
    fps: float
    frameCount: int
    def decodeFrame(self, index: int, outFrame: Frame) -> None: pass
    def getAnimationMatrices(self, frameCache: FrameCache, index: int | float, skeleton: ISkeleton) -> np.ndarray: pass

# FrameTuple
class FrameTuple(NamedTuple):
    frameIndex: int
    frame: Frame

# FrameCache
class FrameCache:
    frameFactory: callable = lambda skeleton: Frame(skeleton)

    previousFrame: FrameTuple #(int, Frame)
    nextFrame: FrameTuple #(int, Frame)
    interpolatedFrame: Frame
    skeleton: ISkeleton

    def __init__(self, skeleton: ISkeleton):
        self.previousFrame = FrameTuple(-1, FrameCache.frameFactory(skeleton))
        self.nextFrame = FrameTuple(-1, FrameCache.frameFactory(skeleton))
        self.interpolatedFrame = FrameCache.frameFactory(skeleton)
        self.skeleton = skeleton
        self.clear()

    def clear(self) -> None:
        self.previousFrame = FrameTuple(-1, self.previousFrame.frame); self.previousFrame.frame.clear(self.skeleton)
        self.nextFrame = FrameTuple(-1, self.nextFrame.frame); self.nextFrame.frame.clear(self.skeleton)

    def getFrame(self, anim: IAnimation, index: int | float) -> Frame:
        match index:
            case time if isinstance(index, float):
                # calculate the index of the current frame
                frameIndex = int(time * anim.fps) % anim.frameCount
                t = (time * anim.fps - frameIndex) % 1
                # get current and next frame
                frame1 = self.getFrame(anim, frameIndex)
                frame2 = self.getFrame(anim, (frameIndex + 1) % anim.frameCount)
                # interpolate bone positions, angles and scale
                for i in range(len(frame1.bones)):
                    frame1Bone = frame1.bones[i]
                    frame2Bone = frame2.bones[i]
                    self.interpolatedFrame.bones[i].position = np.interp(t, frame1Bone.position, frame2Bone.position)
                    self.interpolatedFrame.bones[i].angle = quat.slerp_evaluate(frame1Bone.angle, frame2Bone.angle, t)
                    self.interpolatedFrame.bones[i].scale = frame1Bone.scale + (frame2Bone.scale - frame1Bone.scale) * t
                return self.interpolatedFrame
            case frameIndex if isinstance(index, int):
                # try to lookup cached (precomputed) frame - happens when GUI Autoplay runs faster than animation FPS
                if frameIndex == self.previousFrame.frameIndex: return self.previousFrame.frame
                elif frameIndex == self.nextFrame.frameIndex: return self.nextFrame.frame
                # only two frames are cached at a time to minimize memory usage, especially with Autoplay enabled
                frame: Frame
                if frameIndex > self.previousFrame.frameIndex: frame = self.previousFrame.frame; self.previousFrame = self.nextFrame; self.nextFrame = FrameTuple(frameIndex, frame)
                else: frame = self.nextFrame.frame; self.nextFrame = self.previousFrame; self.previousFrame = FrameTuple(frameIndex, frame)
                # we make an assumption that frames within one animation contain identical bone sets, so we don't clear frame here
                anim.decodeFrame(frameIndex, frame)
                return frame
            case _: raise Exception(f'Unknown {data}')

# AnimationController
class AnimationController:
    frameCache: FrameCache
    updateHandler: callable = lambda a, b: None
    activeAnimation: IAnimation
    time: float
    shouldUpdate: bool
    isPaused: bool
    
    @property
    def frame(self) -> int:
        return round(self.time * self.activeAnimation.fps) % self.activeAnimation.frameCount if self.activeAnimation and self.activeAnimation.frameCount != 0 else 0
    @frame.setter
    def setFrame(self, value: int) -> None:
        if self.activeAnimation:
            self.time = value / self.activeAnimation.fps if self.activeAnimation.fps != 0 else 0.
            self.shouldUpdate = True

    def __init__(self, skeleton: ISkeleton):
        self.frameCache = FrameCache(skeleton)

    def update(self, timeStep: float) -> bool:
        if not self.activeAnimation: return False
        if self.isPaused: res = self.shouldUpdate; self.shouldUpdate = False; return res
        self.time += timeStep
        self.updateHandler(self.activeAnimation, self.frame)
        self.shouldUpdate = False
        return True

    def setAnimation(self, animation: IAnimation) -> None:
        self.frameCache.clear()
        self.activeAnimation = animation
        self.time = 0.
        self.updateHandler(self.activeAnimation, -1)

    def pauseLastFrame(self) -> None:
        self.isPaused = True
        self.frame = 0 if not self.activeAnimation else self.activeAnimation.frameCount - 1

    def getAnimationMatrices(self, skeleton: ISkeleton) -> np.ndarray:
        return self.activeAnimation.getAnimationMatrices(self.frameCache, self.frame, skeleton) if self.isPaused else \
            self.activeAnimation.getAnimationMatrices(self.frameCache, self.time, skeleton)

    def registerUpdateHandler(self, handler: callable) -> None: self.updateHandler = handler

#endregion

File: dotnet/Gfx/OpenStack.Gfx.Egin/Egin_Animate.cs
using System;
using System.Collections.Generic;
using System.Numerics;

namespace OpenStack.Gfx.Egin;

#region Animate

/// <summary>
/// Bone
/// </summary>
public class Bone {
    public int Index;
    public Bone Parent;
    public List<Bone> Children = [];
    public string Name;
    public Vector3 Position;
    public Quaternion Angle;
    public Matrix4x4 BindPose;
    public Matrix4x4 InverseBindPose;

    public Bone(int index, string name, Vector3 position, Quaternion rotation) {
        Index = index;
        Name = name;
        Position = position;
        Angle = rotation;
        // calculate matrices
        BindPose = Matrix4x4.CreateFromQuaternion(rotation) * Matrix4x4.CreateTranslation(position);
        Matrix4x4.Invert(BindPose, out InverseBindPose);
    }

    public void SetParent(Bone parent) {
        if (!Children.Contains(parent)) {
            Parent = parent;
            parent.Children.Add(this);
        }
    }
}

/// <summary>
/// ISkeleton
/// </summary>
public interface ISkeleton {
    Bone[] Roots { get; }
    Bone[] Bones { get; }
}

/// <summary>
/// ChannelAttribute
/// </summary>
public enum ChannelAttribute {
    Position = 0,
    Angle = 1,
    Scale = 2,
    Unknown = 3,
}

/// <summary>
/// Frame
/// </summary>
public struct FrameBone {
    public Vector3 Position;
    public Quaternion Angle;
    public float Scale;
}

/// <summary>
/// Frame
/// </summary>
public class Frame {
    public FrameBone[] Bones;

    public Frame(ISkeleton skeleton) {
        Bones = new FrameBone[skeleton.Bones.Length];
        Clear(skeleton);
    }

    public void SetAttribute(int bone, ChannelAttribute attribute, Vector3 data) {
        switch (attribute) {
            case ChannelAttribute.Position: Bones[bone].Position = data; break;
#if DEBUG
            default: Console.WriteLine($"Unknown frame attribute '{attribute}' encountered with Vector3 data"); break;
#endif
        }
    }

    public void SetAttribute(int bone, ChannelAttribute attribute, Quaternion data) {
        switch (attribute) {
            case ChannelAttribute.Angle: Bones[bone].Angle = data; break;
#if DEBUG
            default: Console.WriteLine($"Unknown frame attribute '{attribute}' encountered with Quaternion data"); break;
#endif
        }
    }

    public void SetAttribute(int bone, ChannelAttribute attribute, float data) {
        switch (attribute) {
            case ChannelAttribute.Scale: Bones[bone].Scale = data; break;
#if DEBUG
            default: Console.WriteLine($"Unknown frame attribute '{attribute}' encountered with float data"); break;
#endif
        }
    }

    /// <summary>
    /// Resets frame bones to their bind pose. Should be used on animation change.
    /// </summary>
    /// <param name="skeleton">The same skeleton that was passed to the constructor.</param>
    public void Clear(ISkeleton skeleton) {
        for (var i = 0; i < Bones.Length; i++) {
            Bones[i].Position = skeleton.Bones[i].Position;
            Bones[i].Angle = skeleton.Bones[i].Angle;
            Bones[i].Scale = 1;
        }
    }
}

/// <summary>
/// IAnimation
/// </summary>
public interface IAnimation {
    string Name { get; }
    float Fps { get; }
    int FrameCount { get; }
    void DecodeFrame(int frameIndex, Frame outFrame);
    Matrix4x4[] GetAnimationMatrices(FrameCache frameCache, object index, ISkeleton skeleton);
}

/// <summary>
/// FrameCache
/// </summary>
public class FrameCache {
    public static Func<ISkeleton, Frame> FrameFactory = skeleton => new Frame(skeleton);

    internal (int frameIndex, Frame frame) PreviousFrame;
    internal (int frameIndex, Frame frame) NextFrame;
    internal Frame InterpolatedFrame;
    internal ISkeleton Skeleton;

    public FrameCache(ISkeleton skeleton) {
        PreviousFrame = (-1, FrameFactory(skeleton));
        NextFrame = (-1, FrameFactory(skeleton));
        InterpolatedFrame = FrameFactory(skeleton);
        Skeleton = skeleton;
        Clear();
    }

    /// <summary>
    /// Clears interpolated frame bones and frame cache. Should be used on animation change.
    /// </summary>
    public void Clear() {
        PreviousFrame = (-1, PreviousFrame.frame); PreviousFrame.frame.Clear(Skeleton);
        NextFrame = (-1, NextFrame.frame); NextFrame.frame.Clear(Skeleton);
    }

    /// <summary>
    /// Get the animation frame at a time/index.
    /// </summary>
    /// <param name="time">The time to get the frame for.</param>
    public Frame GetFrame(IAnimation anim, object index) {
        switch (index) {
            case float time: {
                    // calculate the index of the current frame
                    var frameIndex = (int)(time * anim.Fps) % anim.FrameCount;
                    var t = (time * anim.Fps - frameIndex) % 1;
                    // get current and next frame
                    var frame1 = GetFrame(anim, frameIndex);
                    var frame2 = GetFrame(anim, (frameIndex + 1) % anim.FrameCount);
                    // interpolate bone positions, angles and scale
                    for (var i = 0; i < frame1.Bones.Length; i++) {
                        var frame1Bone = frame1.Bones[i];
                        var frame2Bone = frame2.Bones[i];
                        InterpolatedFrame.Bones[i].Position = Vector3.Lerp(frame1Bone.Position, frame2Bone.Position, t);
                        InterpolatedFrame.Bones[i].Angle = Quaternion.Slerp(frame1Bone.Angle, frame2Bone.Angle, t);
                        InterpolatedFrame.Bones[i].Scale = frame1Bone.Scale + (frame2Bone.Scale - frame1Bone.Scale) * t;
                    }
                    return InterpolatedFrame;
                }
            case int frameIndex: {
                    // try to lookup cached (precomputed) frame - happens when GUI Autoplay runs faster than animation FPS
                    if (frameIndex == PreviousFrame.frameIndex) return PreviousFrame.frame;
                    else if (frameIndex == NextFrame.frameIndex) return NextFrame.frame;
                    // only two frames are cached at a time to minimize memory usage, especially with Autoplay enabled
                    Frame frame;
                    if (frameIndex > PreviousFrame.frameIndex) { frame = PreviousFrame.frame; PreviousFrame = NextFrame; NextFrame = (frameIndex, frame); }
                    else { frame = NextFrame.frame; NextFrame = PreviousFrame; PreviousFrame = (frameIndex, frame); }
                    // we make an assumption that frames within one animation contain identical bone sets, so we don't clear frame here
                    anim.DecodeFrame(frameIndex, frame);
                    return frame;
                }
            default: throw new ArgumentOutOfRangeException(nameof(index));
        }
    }
}

/// <summary>
/// AnimationController
/// </summary>
public class AnimationController {
    internal FrameCache FrameCache;
    Action<IAnimation, int> UpdateHandler = (a, b) => { };
    public IAnimation ActiveAnimation;
    float Time;
    bool ShouldUpdate;
    public bool IsPaused;

    public int Frame {
        get => ActiveAnimation != null && ActiveAnimation.FrameCount != 0
            ? (int)Math.Round(Time * ActiveAnimation.Fps) % ActiveAnimation.FrameCount
            : 0;
        set {
            if (ActiveAnimation != null) {
                Time = ActiveAnimation.Fps != 0
                    ? value / ActiveAnimation.Fps
                    : 0f;
                ShouldUpdate = true;
            }
        }
    }

    public AnimationController(ISkeleton skeleton) => FrameCache = new FrameCache(skeleton);

    public bool Update(float timeStep) {
        if (ActiveAnimation == null) return false;
        if (IsPaused) { var res = ShouldUpdate; ShouldUpdate = false; return res; }
        Time += timeStep;
        UpdateHandler(ActiveAnimation, Frame);
        ShouldUpdate = false;
        return true;
    }

    public void SetAnimation(IAnimation animation) {
        FrameCache.Clear();
        ActiveAnimation = animation;
        Time = 0f;
        UpdateHandler(ActiveAnimation, -1);
    }

    public void PauseLastFrame() {
        IsPaused = true;
        Frame = ActiveAnimation == null ? 0 : ActiveAnimation.FrameCount - 1;
    }

    public Matrix4x4[] GetAnimationMatrices(ISkeleton skeleton)
        => IsPaused
        ? ActiveAnimation.GetAnimationMatrices(FrameCache, Frame, skeleton)
        : ActiveAnimation.GetAnimationMatrices(FrameCache, Time, skeleton);

    public void RegisterUpdateHandler(Action<IAnimation, int> handler) => UpdateHandler = handler;
}

#endregion